File: modules/pdf-engine/android/src/main/java/expo/modules/pdfengine/ImagePdfConverter.kt
package expo.modules.pdfengine

import android.content.Context
import android.graphics.*
import android.graphics.pdf.PdfDocument
import android.media.ExifInterface
import android.net.Uri
import android.os.Build
import expo.modules.kotlin.Promise
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File
import java.util.concurrent.Executors
import kotlin.math.*

class ImagePdfOptions : Record {
  @Field var jobId: String = ""
  @Field var uris: List<String> = emptyList()
  @Field var outputUri: String = ""
  @Field var pageSize: String = "a4"
}

private class ConversionFailure(val code: String, message: String) : Exception(message)

class ImagePdfConverter {
  private val worker = Executors.newSingleThreadExecutor()
  private val jobs = PdfJobRegistry()
  fun cancel(id: String) { jobs.cancel(id) }
  fun destroy() { jobs.destroy(worker) }

  fun start(context: Context, options: ImagePdfOptions, promise: Promise, progress: (Int, Int) -> Unit) {
    jobs.submit(worker, options.jobId, promise) { cancelled ->
      var partial: File? = null
      var output: File? = null
      var committed = false
      fun checkCancelled() { if (cancelled.get()) throw ConversionFailure("PDF_CANCELLED", "Conversion cancelled.") }
      try {
        if (options.uris.isEmpty() || options.uris.size > 30 || options.pageSize !in listOf("a4", "letter", "image")) {
          throw ConversionFailure("PDF_INVALID_OPTIONS", "Choose between 1 and 30 images and a valid page size.")
        }
        val uri = Uri.parse(options.outputUri)
        if (uri.scheme != "file") throw ConversionFailure("PDF_INVALID_OUTPUT", "Invalid output location.")
        output = File(requireNotNull(uri.path)).canonicalFile
        val root = File(context.filesDir, "Versara PDFs").canonicalFile
        if (output.parentFile != root || output.exists()) throw ConversionFailure("PDF_INVALID_OUTPUT", "Choose a new PDF output location.")
        root.mkdirs()
        val temporary = File(root, output.name + ".partial")
        if (temporary.exists()) throw ConversionFailure("PDF_INVALID_OUTPUT", "This output is already being created.")
        partial = temporary
        val pixelBudget = min(3_000_000, 24_000_000 / options.uris.size)
        val pdf = PdfDocument()
        try {
          options.uris.forEachIndexed { index, source ->
            checkCancelled()
            val bitmap = decode(context, source, pixelBudget)
            try {
              checkCancelled()
              var w = if (options.pageSize == "letter") 612 else 595
              var h = if (options.pageSize == "letter") 792 else 842
              val margin = if (options.pageSize == "image") 0f else 18f
              if (options.pageSize == "image") {
                val ratio = 842f / max(bitmap.width, bitmap.height)
                w = max(1, (bitmap.width * ratio).roundToInt())
                h = max(1, (bitmap.height * ratio).roundToInt())
              } else if (bitmap.width > bitmap.height) { val swap = w; w = h; h = swap }
              val page = pdf.startPage(PdfDocument.PageInfo.Builder(w, h, index + 1).create())
              try {
                page.canvas.drawColor(Color.WHITE)
                val scale = min((w - margin * 2) / bitmap.width, (h - margin * 2) / bitmap.height)
                val dw = bitmap.width * scale
                val dh = bitmap.height * scale
                val rect = RectF((w - dw) / 2, (h - dh) / 2, (w + dw) / 2, (h + dh) / 2)
                page.canvas.drawBitmap(bitmap, null, rect, Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG))
              } finally { pdf.finishPage(page) }
            } finally { bitmap.recycle() }
            progress(index + 1, options.uris.size)
          }
          checkCancelled()
          partial.outputStream().use { pdf.writeTo(it) }
        } finally { pdf.close() }
        checkCancelled()
        if (partial.length() == 0L || !partial.renameTo(output)) throw ConversionFailure("PDF_WRITE_FAILED", "Could not save the PDF. Check available storage.")
        committed = true
        checkCancelled()
        promise.resolve(mapOf("uri" to Uri.fromFile(output).toString(), "pageCount" to options.uris.size, "size" to output.length().toDouble()))
      } catch (error: ConversionFailure) {
        if (committed) output?.delete()
        promise.reject(error.code, error.message, error)
      } catch (error: OutOfMemoryError) {
        if (committed) output?.delete()
        promise.reject("PDF_MEMORY_LIMIT", "Not enough memory. Try fewer images.", null)
      } catch (error: Exception) {
        if (committed) output?.delete()
        promise.reject("PDF_CONVERSION_FAILED", "Could not convert an image. Use readable JPG, PNG or supported HEIC files and check free storage.", error)
      } finally { partial?.delete() }
    }
  }

  private fun decode(context: Context, value: String, budget: Int): Bitmap {
    val uri = Uri.parse(value)
    if (uri.scheme != "file") throw ConversionFailure("PDF_INVALID_IMAGE", "Select images using the file browser.")
    val file = File(requireNotNull(uri.path)).canonicalFile
    if (!file.path.startsWith(context.cacheDir.canonicalPath + File.separator)) throw ConversionFailure("PDF_INVALID_IMAGE", "Select images using the file browser.")
    if (Build.VERSION.SDK_INT >= 28) {
      return ImageDecoder.decodeBitmap(ImageDecoder.createSource(file)) { decoder, info, _ ->
        decoder.allocator = ImageDecoder.ALLOCATOR_SOFTWARE
        val scale = min(1.0, min(2200.0 / max(info.size.width, info.size.height), sqrt(budget.toDouble() / (info.size.width.toDouble() * info.size.height))))
        decoder.setTargetSize(max(1, (info.size.width * scale).toInt()), max(1, (info.size.height * scale).toInt()))
      }
    }
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.path, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) throw ConversionFailure("PDF_IMAGE_UNSUPPORTED", "This device cannot read this image. Try JPG or PNG.")
    var sample = 1
    while (bounds.outWidth.toDouble() * bounds.outHeight / sample / sample > budget || max(bounds.outWidth, bounds.outHeight) / sample > 2200) sample *= 2
    val bitmap = BitmapFactory.decodeFile(file.path, BitmapFactory.Options().apply { inSampleSize = sample; inPreferredConfig = Bitmap.Config.ARGB_8888 })
      ?: throw ConversionFailure("PDF_IMAGE_UNSUPPORTED", "This image is unreadable. Try another image.")
    val orientation = try { ExifInterface(file.path).getAttributeInt(ExifInterface.TAG_ORIENTATION, 1) } catch (_: Exception) { 1 }
    val matrix = Matrix()
    when (orientation) {
      2 -> matrix.setScale(-1f, 1f)
      3 -> matrix.setRotate(180f)
      4 -> matrix.setScale(1f, -1f)
      5 -> { matrix.setRotate(90f); matrix.postScale(-1f, 1f) }
      6 -> matrix.setRotate(90f)
      7 -> { matrix.setRotate(-90f); matrix.postScale(-1f, 1f) }
      8 -> matrix.setRotate(-90f)
    }
    if (matrix.isIdentity) return bitmap
    return try { Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true) } finally { bitmap.recycle() }
  }
}

File: modules/pdf-engine/ios/ImagePdfConverter.swift
import ExpoModulesCore
import CoreGraphics
import ImageIO
import Foundation

struct ImagePdfOptions: Record {
  @Field var jobId = ""
  @Field var uris: [String] = []
  @Field var outputUri = ""
  @Field var pageSize = "a4"
}

private struct ConversionFailure: Error {
  let code: String
  let message: String
}

final class ImagePdfConverter {
  private let queue = DispatchQueue(label: "com.versara.pdf.convert", qos: .userInitiated)
  private let lock = NSLock()
  private var cancelled = Set<String>()
  private var active = Set<String>()
  private var destroyed = false

  func cancel(_ id: String) {
    lock.lock(); defer { lock.unlock() }
    guard !destroyed else { return }
    if cancelled.count >= 64, let stale = cancelled.first(where: { !active.contains($0) }) { cancelled.remove(stale) }
    cancelled.insert(id)
  }
  func destroy() { lock.lock(); destroyed = true; cancelled.removeAll(); lock.unlock() }
  private func check(_ id: String) throws {
    lock.lock()
    let shouldCancel = destroyed || cancelled.contains(id)
    lock.unlock()
    if shouldCancel { throw ConversionFailure(code: "PDF_CANCELLED", message: "Conversion cancelled.") }
  }

  func start(_ options: ImagePdfOptions, promise: Promise, progress: @escaping (Int, Int) -> Void) {
    lock.lock()
    guard !destroyed, active.count < 4, !active.contains(options.jobId) else {
      lock.unlock()
      promise.reject("PDF_BUSY", "Another PDF task is finishing or this tool has closed. Please try again.")
      return
    }
    active.insert(options.jobId); lock.unlock()
    queue.async {
      var partial: URL?
      var output: URL?
      var committed = false
      defer {
        if let partial { try? FileManager.default.removeItem(at: partial) }
        self.lock.lock(); self.cancelled.remove(options.jobId); self.active.remove(options.jobId); self.lock.unlock()
      }
      do {
        guard (1...30).contains(options.uris.count), ["a4", "letter", "image"].contains(options.pageSize) else {
          throw ConversionFailure(code: "PDF_INVALID_OPTIONS", message: "Choose between 1 and 30 images and a valid page size.")
        }
        let root = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent("Versara PDFs", isDirectory: true)
        guard let destination = URL(string: options.outputUri), destination.isFileURL,
          destination.deletingLastPathComponent().resolvingSymlinksInPath().standardizedFileURL.path == root.resolvingSymlinksInPath().standardizedFileURL.path,
          !FileManager.default.fileExists(atPath: destination.path) else {
          throw ConversionFailure(code: "PDF_INVALID_OUTPUT", message: "Choose a new PDF output location.")
        }
        output = destination
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        let temporary = destination.appendingPathExtension("partial")
        guard !FileManager.default.fileExists(atPath: temporary.path) else {
          throw ConversionFailure(code: "PDF_INVALID_OUTPUT", message: "This output is already being created.")
        }
        partial = temporary
        try self.check(options.jobId)
        guard let context = CGContext(temporary as CFURL, mediaBox: nil, nil) else {
          throw ConversionFailure(code: "PDF_WRITE_FAILED", message: "Could not create the PDF. Check free storage.")
        }
        do {
          let budget = min(3_000_000, 24_000_000 / options.uris.count)
          for (index, value) in options.uris.enumerated() {
            try autoreleasepool {
              try self.check(options.jobId)
              let image = try self.decode(value, budget: budget)
              var width: CGFloat = options.pageSize == "letter" ? 612 : 595
              var height: CGFloat = options.pageSize == "letter" ? 792 : 842
              let margin: CGFloat = options.pageSize == "image" ? 0 : 18
              if options.pageSize == "image" {
                let ratio = 842 / CGFloat(max(image.width, image.height))
                width = max(1, (CGFloat(image.width) * ratio).rounded())
                height = max(1, (CGFloat(image.height) * ratio).rounded())
              } else if image.width > image.height { swap(&width, &height) }
              var box = CGRect(x: 0, y: 0, width: width, height: height)
              let data = NSData(bytes: &box, length: MemoryLayout<CGRect>.size)
              context.beginPDFPage([kCGPDFContextMediaBox: data] as CFDictionary)
              context.setFillColor(CGColor(gray: 1, alpha: 1))
              context.fill(box)
              let scale = min((width - margin * 2) / CGFloat(image.width), (height - margin * 2) / CGFloat(image.height))
              let dw = CGFloat(image.width) * scale
              let dh = CGFloat(image.height) * scale
              context.interpolationQuality = .high
              context.draw(image, in: CGRect(x: (width - dw) / 2, y: (height - dh) / 2, width: dw, height: dh))
              context.endPDFPage()
            }
            progress(index + 1, options.uris.count)
          }
          context.closePDF()
        } catch { context.closePDF(); throw error }
        try self.check(options.jobId)
        guard let pdf = CGPDFDocument(temporary as CFURL), pdf.numberOfPages == options.uris.count else {
          throw ConversionFailure(code: "PDF_WRITE_FAILED", message: "Could not finish the PDF. Check free storage.")
        }
        try FileManager.default.moveItem(at: temporary, to: destination)
        committed = true
        try self.check(options.jobId)
        let size = (try FileManager.default.attributesOfItem(atPath: destination.path)[.size] as? NSNumber)?.doubleValue ?? 0
        promise.resolve(["uri": destination.absoluteString, "pageCount": options.uris.count, "size": size])
      } catch {
        if committed, let output { try? FileManager.default.removeItem(at: output) }
        if let failure = error as? ConversionFailure { promise.reject(failure.code, failure.message) }
        else { promise.reject("PDF_CONVERSION_FAILED", "Could not convert the images. Check the files and available storage.") }
      }
    }
  }

  private func decode(_ value: String, budget: Int) throws -> CGImage {
    let cache = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].resolvingSymlinksInPath().path + "/"
    guard let url = URL(string: value), url.isFileURL, url.resolvingSymlinksInPath().path.hasPrefix(cache),
      let source = CGImageSourceCreateWithURL(url as CFURL, [kCGImageSourceShouldCache: false] as CFDictionary),
      let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let w = properties[kCGImagePropertyPixelWidth] as? NSNumber,
      let h = properties[kCGImagePropertyPixelHeight] as? NSNumber, w.doubleValue > 0, h.doubleValue > 0 else {
      throw ConversionFailure(code: "PDF_INVALID_IMAGE", message: "An image is unreadable. Choose JPG, PNG or HEIC files using Browse files.")
    }
    let scale = min(1, sqrt(Double(budget) / (w.doubleValue * h.doubleValue)))
    let edge = max(1, min(2200, Int(max(w.doubleValue, h.doubleValue) * scale)))
    let options: [CFString: Any] = [kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceCreateThumbnailWithTransform: true, kCGImageSourceThumbnailMaxPixelSize: edge,
      kCGImageSourceShouldCacheImmediately: true]
    guard let image = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary) else {
      throw ConversionFailure(code: "PDF_IMAGE_UNSUPPORTED", message: "This image could not be decoded. Try another image.")
    }
    return image
  }
}
